packages/ai/src/workflows/Workflow.ts

```diff
@@ -292,25 +296,43 @@ export class Workflow<Shape extends IOShape = IOShape> {
     return this.getCurrentStep()?.getState().kind !== "PENDING";
   }
 
-  getRecentValidCode(): CodeArtifact | undefined {
-    let code: CodeArtifact | undefined;
+  getRecentStepWithSuccessCode(
+    requireSuccess: boolean
+  ): LLMStepInstance<IOShape, Shape> | undefined {
+    let recentStep: LLMStepInstance<IOShape, Shape> | undefined;
 
     // look for first code output in the last step that generated any code
     for (let i = this.steps.length - 1; i >= 0; i--) {
       const step = this.steps[i];
       const outputs = step.getOutputs();
       for (const output of Object.values(outputs)) {
         if (output?.kind === "code") {
-          code = output;
+          if (!requireSuccess || output.value.type === "success") {
+            recentStep = step;
+          }
         }
       }
-      if (code) break;
+      if (recentStep) break;
+    }
+    return recentStep;
+  }
+
+  getRecentSuccessCode(): CodeArtifact | undefined {
+    const recentStep =
```

Comment:
we prioritize code that successfully runs, but if there is none, will take the most recent code of any status